### ISSUE TITLE
easier way to update parameters on the fly

### ISSUE THREAD

**Issue by MattHJensen** [OpenSourcePolicyCenter/Tax-Calculator#283]
There is not a good way to update tax law parameter on a calculator after it has been initialized. 

`calc.params.II_em += 100` increases the personal exemption for the current calculator year, but that change is reflected when you run calc.increment_year(). The behavior is not entirely user friendly, but it is fairly intuitive *if* you understand the difference between the underscore and no underscore version of the params. 

`calc.params._II_em +=100` (with an underscore) increases the personal exemption for every future year, but not the current year. This is neither user friendly nor particularly intuitive. Even if this command did update the personal exemption for the current year as well as out years, which might be more intuitive, it wouldn't be entirely useful because you can only change all values in relation to the current values by a fixed amount, with `+= or -+`, or you have to write out the entire array by hand. 

I think we should have a method lets you re-"expand" a parameter array, taking into account whether the parameter is cpi-indexed, after you have changed the value for the current year.

What do others think? In particular, @talumbau, @Amy-Xu, @mmessick?

**Comment by feenberg** [comment]
On Wed, 10 Jun 2015, Matt Jensen wrote:

> 
> There is not a good way to update tax law parameter on a calculator after it
> has been initialized.
> 
> calc.params.II_em += 100 increases the personal exemption for the current
> calculator year, but that change is reflected when you run
> calc.increment_year(). The behavior is not entirely user friendly, but it is
> fairly intuitive if you understand the difference between the underscore and
> no underscore version of the params.
> 
> calc.params._II_em +=100 (with an underscore) increases the personal
> exemption for every future year, but not the current year. This is neither
> user friendly nor particularly intuitive. Even if this command did update
> the personal exemption for the current year as well as out years, which
> might be more intuitive, it wouldn't be entirely useful because you can only
> change all values in relation to the current values by a fixed amount, with
> += or -+, or you have to write out the entire array by hand.
> 
> I think we should have a method lets you re-"expand" a parameter array,
> taking into account whether the parameter is cpi-indexed, after you have
> changed the value for the current year.
> 
> What do others think? In particular, @talumbau, @Amy-Xu, @mmessick?

I agree that the the change should precede the expansion.

dan

**Comment by talumbau** [comment]
This is definitely a wart that should be fixed. I'll throw something out there:

```
calc.params.update_current('II_em',  500)
```
This makes the current year's value of `II_em` 500 and does nothing for future years.

```
calc.params.update('II_em',  [500, 600],  force_cpi=True)
```

This updates the current year's value of `II_em` and the next year, and CPI flag is considered 'True' for the years after that.

**Comment by MattHJensen** [comment]
@talumbau, is your thinking that 

    calc.params.update('II_em', [500])

would change the current year's value of `II_em` to 500 and then apply the default CPI behavior after that? 

And something like this would work without additional backend complexity, right?

    calc.params.update('II_em', [calc.params.II_em + 500]) 

An alternative approach might be doing something like what is described below, but I'm thinking that what you have outlined is superior:

1. reset the value of calc.params._II_em, starting from the default year, to where you want it. This is akin to the values in params.json:
    `calc.params._II_em = [500, 600]`

2. separately, change the cpi behavior, if you'd like. 
    `calc.params.II_em.cpi_inflated = False`

3. tell the calculator to re-expand the arrays, just like it did upon initialization
`calc.params.expand`
or
`calc.params.II_em.expand`

**Comment by talumbau** [comment]
You correctly captured my thought, but I have now just thought of an additional complexity. I encountered this before when setting values on a Calculator object for the web app calculations. Let's consider the actual case of `_II_em`. In the current params.json, the values are 3900, 3950, 4000 for the years 2013, 2014, and 2015, respectively. Also, `cpi_inflated` is True. Let's say one creates a Calculator object with a start year of 2013 and then advances the year to 2015:

```
   # Default Plans
    # Create a default Parameters object
    params = Parameters(start_year=2013)
    # Create a Calculator
    calc1 = Calculator(params=params, records=records)
    calc1.increment_year()
    calc1.increment_year()
```
So now, `calc1.II_em` is 4000. `cacl1._II_em` looks like this:

```
>>> calc1.params._II_em
array([3900, 3950, 4000, 4080, 4165, 4257, 4355, 4459, 4566, 4676, 4788,
       4903])
```
Now what should happen with this statement:

```
calc.params.update('II_em', [4100])
```

We definitely want `calc1.II_em == 4100` immediately after this statement. But the likely desired value of `calc1._II_em` is probably:

```
[3900, 3950, 4100, 4100*(1.0+cpi_rate_2016), ...]
```
In other words, we update `_II_em` with respect to the current year of the Calculator object, basically ignoring the values corresponding to years that have gone by. 

Also, I think we would need two flags: `force_cpi` and `force_no_cpi` since by default the behavior would be driven by the value of `cpi_inflated` for that parameter. Do you agree?

**Comment by MattHJensen** [comment]
Yea, I was also thinking about what would happen to the trailing years that the user had incremented past and also came to the conclusion that they should be left alone. 

Is there a reason that we wouldn't just modify `cpi_inflated` directly rather than creating the new `force_cpi` and `force_no_cpi`? It would be confusing to do

    calc.params.update('II_em', [4100], force_no_cpi)

and then have 

    [3900, 3950, 4100, 4100, 4100 ....]

but still have `cpi_inflated=True`.

It seems more intuitive to do

    calc.params.update('II_em', [4100], cpi_inflated=False)

If you didn't want to change cpi_inflated, you would just do

    calc.params.update('II_em', [4100])

and get 

    [3900, 3950, 4100, 4100*(1.0+cpi_rate_2016), ...]

**Comment by Amy-Xu** [comment]
I agree that we do need a third argument in this update function to give users the option of how they want this parameter inflated. But I feel maybe the default value of this third argument, by default I mean if it's not specified, should be the default cpi_inflated value we have in params.json for this parameter, just as what we did for user_mods. In this way, the program would not cpi-inflate any parameter that shouldn't be inflated.

In addition to this default value, the users should also have the options of their own choice on cpi-inflation. In terms of naming convention, I'm fine with either way. I guess under most economic assumptions, cpi-inflation is a very natural process for most parameters. So it doesn't sound like a 'forced' process. But from the programming perspective, we need to impose cpi rates on parameters, otherwise those parameters don't know how they should grow.

**Comment by MattHJensen** [comment]
resolved by implement_reform()
